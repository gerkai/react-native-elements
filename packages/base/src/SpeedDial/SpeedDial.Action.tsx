import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { FAB, FABProps } from '../FAB/index';
import { RneFunctionComponent } from '../helpers';

export interface SpeedDialActionProps extends Omit<FABProps, 'size'> {}

/** Adds Action to the SpeedDial.
 * This, Receive all [Fab](fab#props) props. */
export const SpeedDialAction: RneFunctionComponent<SpeedDialActionProps> = ({
  title,
  titleStyle,
  placement,
  ...actionProps
}) => {
  return (
    <View
      style={[
        styles.action,
        placement === 'left' && {
          flexDirection: 'row-reverse',
        },
      ]}
    >
      {title && <Text style={[styles.title, titleStyle]}>{title}</Text>}
      <FAB {...actionProps} size="small" style={[actionProps.style]} />
    </View>
  );
};

const styles = StyleSheet.create({
  title: {
    backgroundColor: 'white',
    color: 'black',
    borderRadius: 5,
    paddingHorizontal: 12,
    paddingVertical: 6,
    marginVertical: 8,
    marginHorizontal: 16,
    elevation: 2,
  },
  action: {
    marginBottom: 16,
    marginRight: 24,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
});

SpeedDialAction.displayName = 'SpeedDial.Action';
