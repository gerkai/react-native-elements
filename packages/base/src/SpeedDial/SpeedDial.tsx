import React from 'react';
import {
  View,
  Animated,
  StyleSheet,
  SafeAreaView,
  Pressable,
  PressableProps,
} from 'react-native';
import { FAB, FABProps } from '../FAB';
import { IconNode } from '../Icon';
import Color from 'color';
import { defaultTheme, RneFunctionComponent } from '../helpers';

export interface SpeedDialProps extends FABProps {
  /** Opens the action stack. */
  isOpen: boolean;

  /** Callback fired when the component requests to be open. */
  onOpen?: () => void;

  /** Callback fired when the component requests to be closed. */
  onClose?: () => void;

  /** Icon shown on FAB when action stack is open. */
  openIcon?: IconNode;

  /** Add overlay color to the speed dial. */
  overlayColor?: string;

  /** SpeedDial Action as children. */
  children?: React.ReactChild[];

  /** The duration for the transition, in milliseconds. */
  transitionDuration?: number;

  /** Props for Backdrop Pressable */
  backdropPressableProps?: PressableProps;
}

/** When pressed, a floating action button can display three to six related actions in the form of a speed dial.
 * If more than six actions are needed, something other than a FAB should be used to present them.
 * Upon press, the FAB remains visible and emits a stack of related actions.
 * If the FAB is tapped in this state, it should either initiate its default action or close the speed dial actions.
 */
export const SpeedDial: RneFunctionComponent<SpeedDialProps> = ({
  isOpen,
  onOpen = () => {},
  onClose = () => {},
  icon,
  openIcon,
  children,
  transitionDuration = 150,
  style,
  overlayColor,
  theme = defaultTheme,
  backdropPressableProps: pressableProps,
  ...rest
}) => {
  const animations = React.useRef<Animated.Value[]>(
    [...new Array(React.Children.count(children))].map(
      () => new Animated.Value(Number(isOpen))
    )
  );

  React.useEffect(() => {
    Animated.stagger(
      50,
      animations.current
        .map((animation) =>
          Animated.timing(animation, {
            toValue: Number(isOpen),
            duration: transitionDuration,
            useNativeDriver: true,
          })
        )
        [isOpen ? 'reverse' : 'sort']()
    ).start();
  }, [isOpen, animations, children, transitionDuration]);

  return (
    <View style={[styles.container, style]} pointerEvents="box-none">
      {/* For overlay  */}
      <Pressable
        {...pressableProps}
        onPress={onClose}
        style={[StyleSheet.absoluteFillObject]}
        pointerEvents={isOpen ? 'auto' : 'none'}
      >
        <Animated.View
          style={[
            StyleSheet.absoluteFillObject,
            {
              opacity: animations.current[0],
              backgroundColor:
                overlayColor ||
                Color(theme?.colors?.black).alpha(0.6).rgb().toString(),
            },
          ]}
        />
      </Pressable>

      <SafeAreaView pointerEvents="box-none" style={styles.safeArea}>
        {React.Children.toArray(children).map((ChildAction, i: number) => (
          <Animated.View
            pointerEvents={isOpen ? 'auto' : 'none'}
            key={i}
            style={{
              transform: [{ scale: animations.current[i] }],
              opacity: animations.current[i],
            }}
          >
            {ChildAction}
          </Animated.View>
        ))}
        <FAB
          style={[styles.fab]}
          icon={isOpen ? openIcon : icon}
          theme={theme}
          {...rest}
          onPress={isOpen ? onClose : onOpen}
        />
      </SafeAreaView>
    </View>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    alignItems: 'flex-end',
  },
  container: {
    ...StyleSheet.absoluteFillObject,
    justifyContent: 'flex-end',
  },
  fab: {
    margin: 16,
    marginTop: 0,
  },
});

SpeedDial.displayName = 'SpeedDial';
