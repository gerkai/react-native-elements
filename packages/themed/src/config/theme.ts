import { Colors } from './colors';
import { ComponentTheme } from './theme.component';

export type RecursivePartial<T> = { [P in keyof T]?: RecursivePartial<T[P]> };

export type ThemeMode = 'light' | 'dark';

export interface ThemeSpacing {
  xs: number;
  sm: number;
  md: number;
  lg: number;
  xl: number;
}

export interface Theme {
  mode: ThemeMode;
  spacing: ThemeSpacing;
}

export interface FullTheme extends ComponentTheme, Theme {
  colors: Colors;
}
