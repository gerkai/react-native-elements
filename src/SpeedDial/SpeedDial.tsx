import React from 'react';
import {
  View,
  Animated,
  StyleSheet,
  SafeAreaView,
  TouchableWithoutFeedback,
} from 'react-native';
import FAB, { FABProps } from '../FAB';
import { IconNode } from '../Icon';
import Color from 'color';
import { RneFunctionComponent } from '../helpers';

export type SpeedDialProps = {
  isOpen: boolean;
  onOpen: () => void;
  onClose: () => void;
  openIcon?: IconNode;
  overlayColor?: string;
  children?: React.ReactChild[];
  transitionDuration?: number;
} & FABProps;

export const SpeedDial: RneFunctionComponent<SpeedDialProps> = ({
  isOpen,
  onOpen = () => {},
  onClose = () => {},
  icon,
  openIcon,
  children,
  transitionDuration = 150,
  style,
  overlayColor,
  theme,
  ...props
}) => {
  const animations = React.useRef<Animated.Value[]>(
    [...new Array(React.Children.count(children))].map(
      () => new Animated.Value(Number(isOpen))
    )
  );

  React.useEffect(() => {
    Animated.stagger(
      50,
      animations.current
        .map((animation) =>
          Animated.timing(animation, {
            toValue: Number(isOpen),
            duration: transitionDuration,
            useNativeDriver: true,
          })
        )
        [isOpen ? 'reverse' : 'sort']()
    ).start();
  }, [isOpen, animations, children, transitionDuration]);

  return (
    <View style={[styles.container, style]} pointerEvents="box-none">
      <TouchableWithoutFeedback onPress={onClose}>
        <Animated.View
          style={[
            StyleSheet.absoluteFillObject,
            {
              opacity: animations.current[0],
              backgroundColor:
                overlayColor ||
                Color(theme?.colors?.black).alpha(0.6).rgb().toString(),
            },
          ]}
          pointerEvents={isOpen ? 'auto' : 'none'}
        />
      </TouchableWithoutFeedback>

      <SafeAreaView pointerEvents="box-none" style={styles.safeArea}>
        {React.Children.toArray(children).map((ChildAction, i: number) => (
          <Animated.View
            pointerEvents={isOpen ? 'auto' : 'none'}
            key={i}
            style={{
              transform: [{ scale: animations.current[i] }],
              opacity: animations.current[i],
            }}
          >
            {ChildAction}
          </Animated.View>
        ))}
        <FAB
          style={[styles.fab]}
          icon={isOpen ? openIcon : icon}
          {...props}
          onPress={isOpen ? onClose : onOpen}
        />
      </SafeAreaView>
    </View>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    alignItems: 'flex-end',
  },
  container: {
    ...StyleSheet.absoluteFillObject,
    justifyContent: 'flex-end',
  },
  fab: {
    margin: 16,
    marginTop: 0,
  },
});

SpeedDial.displayName = 'SpeedDial';
