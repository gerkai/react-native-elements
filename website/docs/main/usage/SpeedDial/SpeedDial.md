import Snack from './snack/index.md'

<Snack />

### Child Components

#### SpeedDial.Action

> Receives all [FAB](../../fab.md#props) props.
