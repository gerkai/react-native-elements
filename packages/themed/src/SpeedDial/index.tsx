import { withTheme } from '../config';
import {
  SpeedDial,
  SpeedDialProps,
} from '@react-native-elements/base/dist/SpeedDial/SpeedDial';
import {
  SpeedDialAction,
  SpeedDialActionProps,
} from '@react-native-elements/base/dist/SpeedDial/SpeedDial.Action';

export type { SpeedDialProps, SpeedDialActionProps };
export default Object.assign(withTheme(SpeedDial, 'SpeedDial'), {
  Action: withTheme(SpeedDialAction, 'SpeedDialAction'),
});
