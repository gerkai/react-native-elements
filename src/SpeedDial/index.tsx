import { withTheme } from '../config';
import { SpeedDial, SpeedDialProps } from './SpeedDial';
import {
  SpeedDialAction,
  SpeedDialActionProps,
} from './components/SpeedDialAction';

export type { SpeedDialProps, SpeedDialActionProps };
export default Object.assign(withTheme(SpeedDial, 'SpeedDial'), {
  Action: withTheme(SpeedDialAction, 'SpeedDialAction'),
});
