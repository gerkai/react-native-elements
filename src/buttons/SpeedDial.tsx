import React from 'react';
import {
  Text,
  View,
  Animated,
  StyleSheet,
  SafeAreaView,
  TouchableWithoutFeedback,
} from 'react-native';
import FAB, { FABProps } from './FAB';
import { withTheme } from '../config';
import { IconNode } from '../icons/Icon';
import Color from 'color';
import { RneFunctionComponent } from '../helpers';

export type SpeedDialActionProps = Omit<FABProps, 'size'>;

const SpeedDialAction: RneFunctionComponent<SpeedDialActionProps> = withTheme(
  ({ title, titleStyle, ...actionProps }) => {
    return (
      <View style={styles.action}>
        {title && <Text style={[styles.title, titleStyle]}>{title}</Text>}
        <FAB {...actionProps} size="small" style={[actionProps.style]} />
      </View>
    );
  },
  'SpeedDialAction'
);

export type SpeedDialProps = {
  isOpen: boolean;
  onOpen: () => void;
  onClose: () => void;
  openIcon?: IconNode;
  overlayColor?: string;
  children?: React.ReactChild[];
  transitionDuration?: number;
} & FABProps;

const SpeedDial: RneFunctionComponent<SpeedDialProps> = ({
  theme,
  isOpen,
  onOpen = () => {},
  onClose = () => {},
  icon,
  openIcon,
  children,
  transitionDuration = 150,
  style,
  overlayColor,
  ...props
}) => {
  const animations = React.useRef<Animated.Value[]>(
    [...new Array(React.Children.count(children))].map(
      () => new Animated.Value(Number(isOpen))
    )
  );

  React.useEffect(() => {
    Animated.stagger(
      50,
      animations.current
        .map((animation) =>
          Animated.timing(animation, {
            toValue: Number(isOpen),
            duration: transitionDuration,
            useNativeDriver: true,
          })
        )
        [isOpen ? 'reverse' : 'sort']()
    ).start();
  }, [isOpen, animations, children, transitionDuration]);

  return (
    <View style={[styles.container, style]} pointerEvents="box-none">
      <TouchableWithoutFeedback onPress={onClose}>
        <Animated.View
          style={[
            StyleSheet.absoluteFillObject,
            {
              opacity: animations.current[0],
              backgroundColor:
                overlayColor ||
                Color(theme?.colors?.black).alpha(0.6).rgb().toString(),
            },
          ]}
          pointerEvents={isOpen ? 'auto' : 'none'}
        />
      </TouchableWithoutFeedback>

      <SafeAreaView pointerEvents="box-none" style={styles.safeArea}>
        {React.Children.toArray(children).map((ChildAction, i: number) => (
          <Animated.View
            pointerEvents={isOpen ? 'auto' : 'none'}
            key={i}
            style={{
              transform: [{ scale: animations.current[i] }],
              opacity: animations.current[i],
            }}
          >
            {ChildAction}
          </Animated.View>
        ))}
        <FAB
          style={[styles.fab]}
          icon={isOpen ? openIcon : icon}
          {...props}
          onPress={isOpen ? onClose : onOpen}
        />
      </SafeAreaView>
    </View>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    alignItems: 'flex-end',
  },
  container: {
    ...StyleSheet.absoluteFillObject,
    justifyContent: 'flex-end',
  },
  fab: {
    margin: 16,
    marginTop: 0,
  },
  title: {
    backgroundColor: 'white',
    color: 'black',
    borderRadius: 5,
    paddingHorizontal: 12,
    paddingVertical: 6,
    marginVertical: 8,
    marginHorizontal: 16,
    elevation: 2,
  },
  action: {
    marginBottom: 16,
    marginRight: 24,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
});
export { SpeedDial };

export default Object.assign(withTheme(SpeedDial, 'SpeedDial'), {
  Action: SpeedDialAction,
});
