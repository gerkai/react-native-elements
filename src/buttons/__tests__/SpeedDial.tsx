import React from 'react';
import { SpeedDial } from '../SpeedDial';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import theme from '../../config/theme';

describe('Speed Dial Component', () => {
  beforeEach(() => {
    let useEffect = jest.spyOn(React, 'useEffect');
    useEffect.mockImplementation((f) => f());
    const app = shallow(<SpeedDial open theme={theme} />);
    expect(app.length).toBe(1);
  });
  it('should render without issues', () => {
    const app = shallow(
      <SpeedDial
        theme={theme}
        isOpen={true}
        icon={{ name: 'edit', color: '#fff' }}
        openIcon={{ name: 'close', color: '#fff' }}
      >
        <SpeedDial.Action
          icon={{ name: 'add', color: '#fff' }}
          title="Add"
          onPress={() => console.log('Add Something')}
        />
        <SpeedDial.Action
          icon={{ name: 'delete', color: '#fff' }}
          title="Delete"
          onPress={() => console.log('Delete Something')}
        />
      </SpeedDial>
    );
    expect(app.length).toBe(1);
    expect(toJson(app)).toMatchSnapshot();
  });
});
