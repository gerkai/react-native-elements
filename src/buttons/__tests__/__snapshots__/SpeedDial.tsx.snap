// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Speed Dial Component should render without issues 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "bottom": 0,
        "justifyContent": "flex-end",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <ForwardRef(AnimatedComponentWrapper)
      pointerEvents="auto"
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "backgroundColor": "rgba(36, 36, 36, 0.6)",
            "opacity": 1,
          },
        ]
      }
    />
  </TouchableWithoutFeedback>
  <ForwardRef(SafeAreaView)
    pointerEvents="box-none"
    style={
      Object {
        "alignItems": "flex-end",
      }
    }
  >
    <ForwardRef(AnimatedComponentWrapper)
      key="0"
      pointerEvents="auto"
      style={
        Object {
          "opacity": 1,
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <Component
        icon={
          Object {
            "color": "#fff",
            "name": "add",
          }
        }
        key=".0"
        onPress={[Function]}
        title="Add"
      />
    </ForwardRef(AnimatedComponentWrapper)>
    <ForwardRef(AnimatedComponentWrapper)
      key="1"
      pointerEvents="auto"
      style={
        Object {
          "opacity": 1,
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <Component
        icon={
          Object {
            "color": "#fff",
            "name": "delete",
          }
        }
        key=".1"
        onPress={[Function]}
        title="Delete"
      />
    </ForwardRef(AnimatedComponentWrapper)>
    <Themed.FAB
      icon={
        Object {
          "color": "#fff",
          "name": "close",
        }
      }
      onPress={[Function]}
      style={
        Array [
          Object {
            "margin": 16,
            "marginTop": 0,
          },
        ]
      }
    />
  </ForwardRef(SafeAreaView)>
</View>
`;
